### PR TITLE
Add more documentation about response headers to tutorial.

### LINKED ISSUE #608: Improve documentation of addHeader
Recently in the Twitt-o-sphere, I described some frustration with Response Headers, and how to send multiple headers at once, while working on a new donation site for Haskell.org. This is because I am a huge whiner (and twitter is where I vent).

But I also found the documentation quite poor, honestly! The only example I ever found was in the test suite, of how to add two headers at once - and that's not great! And it's especially nasty because if you ask me, the way you do it isn't immediately obvious.

Here's an example of why. Let's say I want to add one header. Easy enough:

```haskell
type CorsHeader v = Headers '[ Header "Access-Control-Allow-Origin" String ] v

addCors :: String -> v -> CorsHeader v
addCors origin v = addHeader origin v
```

Now, you have to return `CorsHeader` as part of the type, something like:

```haskell
"foo" :> Get '[PlainText] (CorsHeader String)
```

for example. Then do `return (addCors str)` or something.

OK, so now I want to add multiple headers. My initial intuition was that, despite the fact `Headers` uses an HList, I should have threaded the request something like this:

```haskell
type CorsHeader2 v = Headers '[ Header "Access-Control-Allow-Origin" String ] (Headers [ ... ] v)
```

Why would I think this? Well, because in the documentation the type of `addHeader` isn't explained (I'll get to that in a moment) very much. So I got the impression "Well, clearly this is just some kind of data structure wrapping the final return result. So I should wrap multiple times to make it all work". It's like how if I had `Maybe a` and extended that to `Maybe (Maybe a)`.

OK, so under the assumption I keep "nesting" them, I would then do:

```haskell
addCors2 :: String -> String -> v -> CorsHeader2 v
addCors2 domain next v = addHeader domain $ addHeader next v
```

Except this is totally wrong! Because `addHeader` is built on a type class using an `HList`, this is *the correct definition*, but it absolutely *has the wrong type*.

The proper type of `addCors2` would be something like:

```haskell
type CorsHeader2 v = Headers '[ Header "Access-Control-Allow-Origin" String, Header "Access-Control-..." String ] v
```

That is, because the HList contains the list of headers, the type class instances for `addHeader` "folds" over the HList, adding header entries based on the list position.

This took me like +15 minutes to figure out, *most* of it searching for an example (by frantically searching for `addHeader` in every project using Servant). Why? Because the type of `addHeader` is totally impenetrable!

Thus, I really, really think you should go through a few examples and at least *explain* this mechanism. I'd suggest adding an example in the Haddocks, and in the user guide.

I talked with @alpmestan about this privately on IRC, and he admitted this was one of the less clean parts of Servant. I just want to say, that's totally OK! And I apologize for my whiney twitter rants about such good stuff. :) Everything else has been exceptionally easy to grok, despite all the masterful, amazing type-level magic. But this stumbling block was *especially* frustrating, because adding response headers should conceptually be very easy, and it's very vital!

Thus, even if it's not super clean, that just means it's even *more* important to document!

I'd be willing to open a patch with examples from my actual use case, since they're not entirely theoretical.

### PR DESCRIPTION
**Pull request by jkarni**
Fixes #608 

cc: @thoughtpolice

### CODE REVIEW

**Comment by thoughtpolice** [comment]
LGTM!

File: doc/tutorial/Server.lhs [hunk 1]
```diff
@@ -753,7 +753,30 @@ myHandler :: Server MyHandler
 myHandler = return $ addHeader 1797 albert
 ```
 
-Note that the type of `addHeader x` is different than the type of `x`!
+Note that the type of `addHeader x` is different than the type of `x`! More
```
Comment (soenkehahn) [inline doc/tutorial/Server.lhs]: Shouldn't this say that the type of `addHeader headerValue x` is different from `x`?

Comment (soenkehahn) [inline doc/tutorial/Server.lhs]: Maybe `s/More/Added/`?

File: doc/tutorial/Server.lhs [hunk 2]
```diff
@@ -753,7 +753,30 @@ myHandler :: Server MyHandler
...
+headers will appear in the header list:
+
+``` haskell
+type MyHeadfulHandler = Get '[JSON] (Headers '[Header "X-A-Bool" Bool, Header "X-An-Int" Int] User)
+
+myHeadfulHandler :: Server MyHeadfulHandler
+myHeadfulHandler = return $ addHeader True $ addHeader 1797 albert
+```
+
+But what if your handler adds a header *conditional* on some run-time data? If
```
Comment (soenkehahn) [inline doc/tutorial/Server.lhs]: `s/conditional/optionally depending/`?

File: doc/tutorial/Server.lhs [hunk 3]
```diff
@@ -753,7 +753,30 @@ myHandler :: Server MyHandler
...
+
+``` haskell
+type MyMaybeHeaderHandler
+  = Capture "withHeader" Bool :> Get '[JSON] (Headers '[Header "X-An-Int" Int] User)
+
+myMaybeHeaderHandler :: Server MyMaybeHeaderHandler
+myMaybeHeaderHandler x = return $ if x then addHeader 1797 albert
+                                       else noHeader albert
+```
+
```
Comment (soenkehahn) [inline doc/tutorial/Server.lhs]: Maybe use one newline?

File: doc/tutorial/Server.lhs (hunk 1 above)
Comment (jkarni) [inline doc/tutorial/Server.lhs]: Fixed! Also rephrased something else.

File: doc/tutorial/Server.lhs (hunk 2 above)
Comment (jkarni) [inline doc/tutorial/Server.lhs]: Hm, that doesn't sound quite right either. Maybe:

> But what if your handler only *sometimes* add a header?

Comment (soenkehahn) [inline doc/tutorial/Server.lhs]: Yes, better.

File: doc/tutorial/Server.lhs (hunk 3 above)
Comment (jkarni) [inline doc/tutorial/Server.lhs]: Done.

**Pull request merged**